src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs

```diff
@@ -236,6 +301,80 @@ protected static async Task<ImmutableArray<SyntaxToken>> GetIdentifierOrGlobalNa
             };
         }
 
+        protected static async Task<ImmutableArray<FinderLocation>> FindReferencesInDocumentInsideGlobalSuppressionsAsync(
+            Document document,
+            SemanticModel semanticModel,
+            ISyntaxFacts syntaxFacts,
+            Func<string> getDocumentationCommentId,
+            CancellationToken cancellationToken)
+        {
+            var info = await SyntaxTreeIndex.GetIndexAsync(document, cancellationToken).ConfigureAwait(false);
+            if (!info.ContainsGlobalAttributeInfo)
+            {
+                return ImmutableArray<FinderLocation>.Empty;
+            }
+
+            var suppressMessageAttribute = semanticModel.Compilation.SuppressMessageAttributeType();
+            if (suppressMessageAttribute == null)
+            {
+                return ImmutableArray<FinderLocation>.Empty;
+            }
+
+            var documentationCommentId = getDocumentationCommentId();
+            if (string.IsNullOrEmpty(documentationCommentId) ||
+                !info.TryGetPositionsInGlobalAttributes(documentationCommentId, out var positions))
+            {
+                return ImmutableArray<FinderLocation>.Empty;
+            }
+
+            var root = await semanticModel.SyntaxTree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+            var locations = ArrayBuilder<FinderLocation>.GetInstance();
+            foreach (var position in positions)
+            {
+                var token = root.FindToken(position);
+                if (!IsCandidate(token, semanticModel, syntaxFacts, suppressMessageAttribute, cancellationToken))
+                {
+                    continue;
+                }
+
+                var valueText = token.ValueText;
+                int length = 0;
+                var offsetIntoValueText = position - token.SpanStart;
+                while (offsetIntoValueText < valueText.Length &&
+                    syntaxFacts.IsIdentifierPartCharacter(valueText[offsetIntoValueText++]))
+                {
+                    length++;
+                }
+
+                var location = Location.Create(root.SyntaxTree, new TextSpan(position + 1, length));
+                var containingStringLocation = token.GetLocation();
+                var referenceLocation = new ReferenceLocation(
+                    document,
+                    location,
+                    containingStringLocation);
+
+                locations.Add(new FinderLocation(token.Parent, referenceLocation));
+            }
```

Comment:
please extract thit out and comment it heavily.

Reply:
Sure, done.